File: atlhyper_web/src/api/ingress.ts
/**
 * Ingress API
 *
 * 后端已完成 model_v2 → model 扁平化转换（含行展开），前端直接使用
 */

import { get } from "./request";
import type { IngressOverview, IngressDetail, IngressItem } from "@/types/cluster";

// ============================================================
// 查询参数类型
// ============================================================

interface IngressListParams {
  cluster_id: string;
  namespace?: string;
  limit?: number;
  offset?: number;
}

// ============================================================
// 响应类型（后端返回已展开的扁平行）
// ============================================================

interface IngressListResponse {
  message: string;
  data: IngressItem[];
  total: number;
}

// ============================================================
// API 方法
// ============================================================

/**
 * 获取 Ingress 列表（已按 host×path 展开为多行）
 * GET /api/v2/ingresses?cluster_id=xxx&namespace=xxx
 */
export function getIngressList(params: IngressListParams) {
  return get<IngressListResponse>("/api/v2/ingresses", params);
}

/**
 * 获取 Ingress 详情
 * GET /api/v2/ingresses/{name}?cluster_id=xxx&namespace=xxx
 */
export async function getIngressDetail(data: { ClusterID: string; Namespace: string; Name: string }) {
  return get<{ message: string; data: IngressDetail }>(
    `/api/v2/ingresses/${encodeURIComponent(data.Name)}`,
    {
      cluster_id: data.ClusterID,
      namespace: data.Namespace,
    }
  );
}

// ============================================================
// 概览聚合（前端从扁平行列表计算统计卡片）
// ============================================================

/**
 * 获取 Ingress 概览（包含统计卡片和列表）
 */
export async function getIngressOverview(data: { ClusterID: string }) {
  const response = await getIngressList({ cluster_id: data.ClusterID });
  const rows = response.data.data || [];

  const hostsSet = new Set<string>();
  let tlsCerts = 0;
  let totalPaths = 0;
  const ingressNames = new Set<string>();

  for (const row of rows) {
    if (row.host && row.host !== "*") hostsSet.add(row.host);
    if (row.tls) tlsCerts++;
    if (row.path) totalPaths++;
    ingressNames.add(row.name);
  }

  const overview: IngressOverview = {
    cards: {
      totalIngresses: ingressNames.size,
      usedHosts: hostsSet.size,
      tlsCerts,
      totalPaths,
    },
    rows,
  };

  return {
    ...response,
    data: { data: overview },
  };
}

File: atlhyper_master_v2/model/ingress.go
// atlhyper_master_v2/model/ingress.go
// Ingress Web API 响应类型（camelCase JSON tag，扁平结构）
package model

// IngressItem Ingress 列表项（扁平）
type IngressItem struct {
	Name        string `json:"name"`
	Namespace   string `json:"namespace"`
	Host        string `json:"host"`
	Path        string `json:"path"`
	ServiceName string `json:"serviceName"`
	ServicePort string `json:"servicePort"`
	TLS         bool   `json:"tls"`
	CreatedAt   string `json:"createdAt"`
}

// IngressOverviewCards Ingress 概览统计
type IngressOverviewCards struct {
	TotalIngresses int `json:"totalIngresses"`
	UsedHosts      int `json:"usedHosts"`
	TLSCerts       int `json:"tlsCerts"`
	TotalPaths     int `json:"totalPaths"`
}

// IngressOverview Ingress 概览
type IngressOverview struct {
	Cards IngressOverviewCards `json:"cards"`
	Rows  []IngressItem        `json:"rows"`
}

// IngressDetail Ingress 详情（扁平 + 嵌套 spec/status）
type IngressDetail struct {
	// 基本信息
	Name      string `json:"name"`
	Namespace string `json:"namespace"`
	Class     string `json:"class,omitempty"`
	CreatedAt string `json:"createdAt"`
	Age       string `json:"age,omitempty"`

	// 摘要
	Hosts        []string `json:"hosts,omitempty"`
	TLSEnabled   bool     `json:"tlsEnabled"`
	LoadBalancer []string `json:"loadBalancer,omitempty"`

	// 规格与状态（保留原始结构）
	Spec   interface{} `json:"spec,omitempty"`
	Status interface{} `json:"status,omitempty"`

	// 注解
	Annotations map[string]string `json:"annotations,omitempty"`
}
